### PR TITLE
[ML] Autoscaler metric should suggest scaling down to 0 nodes if possible

### PR DESCRIPTION
**Pull request by hendrikmuhs**
fixes an issue in the _new_ autoscaler (serverless) metrics implementation. Ensure it scale down is suggested down to `0` nodes if possible

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingContext.java [hunk 1]
```diff
@@ -48,48 +48,78 @@ class MlAutoscalingContext {
...
-        anomalyDetectionTasks = List.of();
-        snapshotUpgradeTasks = List.of();
-        dataframeAnalyticsTasks = List.of();
-        modelAssignments = Map.of();
-
-        waitingAnomalyJobs = List.of();
-        waitingSnapshotUpgrades = List.of();
-        waitingAnalyticsJobs = List.of();
-        waitingAllocatedModels = List.of();
+        this(List.of(), List.of(), List.of(), Map.of(), List.of(), null);
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingContext.java:51]: all changes in this file are refactorings, required to mock this context object in the added tests

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java [hunk 2]
```diff
@@ -297,7 +297,7 @@ static void getMemoryAndProcessors(
             && perNodeAvailableModelMemoryInBytes > 0
             && extraModelMemoryInBytes == 0
             && extraProcessors == 0
-            && modelMemoryBytesSum < perNodeMemoryInBytes * (osStatsPerNode.size() - 1)
+            && modelMemoryBytesSum <= perNodeMemoryInBytes * (osStatsPerNode.size() - 1)
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java:300]: ^ that's the fix, a simple off by 1 error

**Comment by hendrikmuhs** [comment]
@elasticmachine update branch

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by hendrikmuhs** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Pull request merged**